<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>10x scATAC-seq processing and analysis: Comparison of chromap vs cellranger-atac using public PBMC and HGMM data</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">chromap_vs_cellranger_scATAC_exploration_10x</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">10x scATAC-seq processing and analysis: Comparison of chromap vs cellranger-atac using public PBMC and HGMM data</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-09-15
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>chromap_vs_cellranger_scATAC_exploration_10x/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong10fdcb0"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> 10fdcb0 </a>
</p>
</div>
<div id="strongRepositoryversionstrong10fdcb0" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version 10fdcb0. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
working directory clean
</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/index.Rmd</code>) and HTML (<code>docs/index.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
10fdcb0
</td>
<td>
jeremymsimon
</td>
<td>
2022-09-14
</td>
<td>
Initial commit
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
112fdea
</td>
<td>
jeremymsimon
</td>
<td>
2022-09-12
</td>
<td>
Start workflowr project.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>This site contains the code and results of my comparative analysis of <a href="https://github.com/haowenz/chromap">chromap v0.1</a> vs <a href="https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/using/count">cellranger-atac v2.0.0</a> in the processing and analysis of 10x scATAC-seq data.</p>
<p>For illustration purposes, I utilize publicly-accessible 10x scATAC-seq data from PBMCs and HGMM (GM12878 B-lymphocytes).</p>
<p>All the code and results of this analysis are available from GitHub at <a href="https://github.com/jeremymsimon/chromap_vs_cellranger_scATAC_exploration_10x" class="uri">https://github.com/jeremymsimon/chromap_vs_cellranger_scATAC_exploration_10x</a>.</p>
<p>Follow the links below to access different pieces of the analysis.</p>
</div>
<div id="motivation" class="section level1">
<h1>Motivation</h1>
<p>My personal preference is to support open-source software whenever possible, while being cognizant of computational efficiency. <code>Cellranger</code> is rather slow (runtimes included in the relevant chapter below), in part because it does more downstream analysis than I desire (clustering, marker gene/region detection, etc). I also like my approaches to remain flexible and modular such that new versions and tools can be easily incorporated as they are introduced.</p>
<p><code>Chromap</code> is one such new approach to align and pre-process scATAC-seq (and other) data very quickly. It by default outputs a fragments file <a href="https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/fragments">akin to cellranger</a>, which contains the chromosome, start, and end coordinates of aligned fragments, along with the cellular barcode associated with that fragment, and the number of read pairs associated with that fragment. <a href="https://www.nature.com/articles/s41467-021-26865-w">The associated manuscript</a> states it is over 10 times (ironically, 10x, lol) faster than traditional approaches e.g. <code>cellranger</code>. Their paper does a great job already of comparing the alignments and resulting clusters to conventional approaches, so that should be your starting point before diving into what I’ve done here. The reason for my duplicating some of those efforts here is because A) their approach utilized <a href="https://liulab-dfci.github.io/MAESTRO/"><code>MAESTRO</code></a> as an intermediate for all processing, and B) they did not utilize a dataset with multiple conditions or replicates. Although <code>MAESTRO</code> is itself a very useful tool, I wished to take either the <code>cellranger</code> or <code>chromap</code> outputs and take a more direct path to R. Here, I utilize the <a href="https://satijalab.org/signac/"><code>Signac</code></a>/<a href="https://satijalab.org/seurat/"><code>Seurat</code></a> framework, but the logic should be generalizable to other ecosystems.</p>
<p>Though to my knowledge this hasn’t been fully demonstrated in the literature, I also wish to incorporate an <a href="https://www.cell.com/cell/fulltext/S0092-8674(21)01279-4">atlas of accessible chromatin from 222 human cell types</a>. The rationale here is that peak-calling on one entire sample at once can miss accessible regions specific to rare cell types or that may be condition specific. As an analogy, for a differential splicing analysis, we may utilize a database of known as well as de novo detected splice junctions unique to our data. This is my thinking here - that we can increase our power to detect important regulatory elements by utilizing genomic loci known to play a regulatory role alongside regions that may be important in our own data. This sort of approach is supported in <code>MAESTRO</code> with <code>--custompeak</code>, so other groups may have a similar philosophy.</p>
<p>There are certainly other ways of aligning scATAC-seq reads, quantifying signal across the genome, and identifying critical regulatory elements for downstream analysis using methods not used or mentioned here. The comparative analysis presented here is not meant to be comprehensive or definitive, nor will I claim my approach is the best or the right-est or the efficient-est.</p>
<p>The analysis presented here merely serves as a proof-of-principle that a <code>cellranger</code>-independent approach is feasible, that <code>chromap</code> seems like a viable alternative, and that this framework can serve as the foundation for future improvements in peak-calling and feature selection.</p>
</div>
<div id="analysis-overview" class="section level1">
<h1>Analysis Overview</h1>
<ol style="list-style-type: decimal">
<li><p><a href="DataDownload.html">scATAC-seq data download</a></p></li>
<li>Running cellranger
<ol style="list-style-type: lower-alpha">
<li><a href="CellrangerPBMC.html">PBMC</a></li>
<li><a href="CellrangerHGMM.html">HGMM</a></li>
</ol></li>
<li>Running chromap
<ol style="list-style-type: lower-alpha">
<li><a href="ChromapPBMC.html">PBMC</a></li>
<li><a href="ChromapHGMM.html">HGMM</a></li>
<li><a href="chromapCompression.html">Compressing fragments files for downstream</a></li>
</ol></li>
<li>Peak-calling on chromap alignments
<ol style="list-style-type: lower-alpha">
<li><a href="MACS2.html">Running MACS2</a></li>
<li><a href="UnionPeaks.html">Creation of union set for downstream analysis</a></li>
</ol></li>
<li>Analysis of chromap data in R
<ol style="list-style-type: lower-alpha">
<li><a href="ChromapSignacInit.html">Basic processing in Signac</a></li>
<li><a href="ChromapSignacQC.html">Compute QC metrics</a></li>
<li><a href="ChromapSignacIntegration.html">Integration, clustering, and estimating Gene Activity</a></li>
</ol></li>
<li>Analysis of cellranger data in R
<ol style="list-style-type: lower-alpha">
<li><a href="CellrangerSignacInitQC.html">Basic processing and QC</a></li>
<li><a href="CellrangerSignacIntegration.html">Integration and clustering</a></li>
</ol></li>
<li>Comparative analyses
<ol style="list-style-type: lower-alpha">
<li><a href="ComparisonPart1.html">Cluster memberships, gene activity, and peak overlaps: approach-specific peaks</a></li>
<li><a href="ComparisonPart2.html">Signal concordance under one unified peak set for both approaches</a></li>
</ol></li>
<li><p><a href="Conclusions.html">Conclusions</a></p></li>
</ol>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
